### ISSUE TITLE
Installation Issues

### ISSUE THREAD

**Issue by dmdu** [dflemin3/approxposterior#21]
I get errors when I try to install the package on a brand new Ubuntu 16.04 instance (run by Docker). 

```
# cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.3 LTS"
# apt-get update
# apt-get install python curl git
# python --version
Python 2.7.12
# curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
# python get-pip.py
# pip —version
pip 10.0.1 from /usr/local/lib/python2.7/dist-packages/pip (python 2.7)

# pip install approxposterior
This fails with the following message:
...
Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-W2oMFN/subprocess32/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-GrV0mc/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-W2oMFN/subprocess32/

Trying to install from source now:
# git clone https://github.com/dflemin3/approxposterior.git
# cd approxposterior/
# python setup.py install
This fails with the following message:
...
Installed /usr/local/lib/python2.7/dist-packages/corner-2.0.1-py2.7.egg
Searching for george
Reading https://pypi.org/simple/george/
Downloading https://files.pythonhosted.org/packages/27/19/9de575be629e3a41c3ca6b1e2c80e0ae90a2e831436c5f70cc8d72e37ab7/george-0.3.1.tar.gz#sha256=175f1a8022430adab8bf8e5e0ffc3941e5f48ca831bd9460e69716cc2820e628
Best match: george 0.3.1
Processing george-0.3.1.tar.gz
Writing /tmp/easy_install-Nz8xxm/george-0.3.1/setup.cfg
Running george-0.3.1/setup.py -q bdist_egg --dist-dir /tmp/easy_install-Nz8xxm/george-0.3.1/egg-dist-tmp-WOjxot
error: Setup script exited with error: SandboxViolation: mkdir('/root/.local', 448) {}
The package setup script has attempted to modify files on your system
that are not within the EasyInstall build area, and has been aborted.
This package cannot be safely installed by EasyInstall, and may not
support alternate installation locations even if you run its setup
script by hand.  Please inform the package's author and the EasyInstall
maintainers to find out if a fix or workaround is available.

Installing conda as recommended in the instructions:
# apt-get install bzip2
# cd /tmp
# curl -O https://repo.continuum.io/archive/Anaconda3-5.0.1-Linux-x86_64.sh
# bash Anaconda3-5.0.1-Linux-x86_64.sh 
# source /root/.bashrc
# conda --version
conda 4.3.30

After conda is installed, pip returns the same error as before:
# pip install approxposterior
...
Command "/root/anaconda3/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-wjqzy3ec/george/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-tmg_y5gf-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-wjqzy3ec/george/

Trying to install the package from source again:
# python setup.py install
...
RuntimeError: Unsupported compiler -- at least C++11 support is needed!

# apt-get install gcc
# gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Same error again:
# python setup.py install
...
RuntimeError: Unsupported compiler -- at least C++11 support is needed!
```

----------------
Please provide more detailed instructions on building the package and avoiding these errors. Please list all dependencies, all required customizations, and all installation steps. If Ubuntu 16.04 is not your primary target, please specify which distribution you want the package to run on.

**Comment by dflemin3** [comment]
It looks like installing the [george](https://github.com/dfm/george) dependency is hanging up both the pip and from source installations.  Our package is now able to be installed using conda, and through some testing, we find that using conda to install approxposterior and its dependencies seems to work well on both target systems, MacOSX and the CentOS linux distribution.  Given this, we have decided to make installation via conda the preferred method and detail those steps below and it the documentation.

**Dependencies**:
The following python packages:
-numpy
-scipy
-scikit-learn
-emcee
-corner
-pytest
-pybind11
-george

**Installation steps**:
Our preferred method for installation is:

`conda install -c conda-forge approxposterior`

where this handles the installation of all dependencies and approxposterior itself.

If installing via pip or from source, the directions are as follows:

Using pip:
```
conda install -c conda-forge george
pip install approxposterior
```

From source:
```
conda install -c conda-forge george
git clone git@github.com:dflemin3/approxposterior.git
cd approxposterior
python setup.py install
```

The final two installation methods involve installing george first to ensure it is properly compiled and working.  I have updated the documentation [on the package website](https://dflemin3.github.io/approxposterior/install.html) and on the package's GitHub [readme](https://github.com/dflemin3/approxposterior/blob/master/README.md) to reflect these changes.

**Comment by dmdu** [comment]
Thanks for looking into this and resolving the issue. `conda install -c conda-forge approxposterior` works fine for me now; installation completes with no error.

I'm looking at something else now while trying to run the example.

```
# conda --version
conda 4.5.6
# conda list | grep approx
approxposterior           0.1                        py_0    conda-forge
# python --version  
Python 3.6.3 :: Anaconda custom (64-bit)

/tmp/approxposterior/examples# python example.py 
Traceback (most recent call last):
  File "example.py", line 15, in <module>
    from approxposterior import bp, likelihood as lh
  File "/root/anaconda3/lib/python3.6/site-packages/approxposterior/__init__.py", line 16, in <module>
    from .bp import *
  File "/root/anaconda3/lib/python3.6/site-packages/approxposterior/bp.py", line 21, in <module>
    from . import plot_utils as pu
  File "/root/anaconda3/lib/python3.6/site-packages/approxposterior/plot_utils.py", line 14, in <module>
    import matplotlib.pyplot as plt
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/pyplot.py", line 113, in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backends/__init__.py", line 60, in pylab_setup
    [backend_name], 0)
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backends/backend_qt5agg.py", line 16, in <module>
    from .backend_qt5 import (
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backends/backend_qt5.py", line 18, in <module>
    import matplotlib.backends.qt_editor.figureoptions as figureoptions
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backends/qt_editor/figureoptions.py", line 20, in <module>
    import matplotlib.backends.qt_editor.formlayout as formlayout
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backends/qt_editor/formlayout.py", line 56, in <module>
    from matplotlib.backends.qt_compat import QtGui, QtWidgets, QtCore
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backends/qt_compat.py", line 137, in <module>
    from PyQt5 import QtCore, QtGui, QtWidgets
ImportError: libGL.so.1: cannot open shared object file: No such file or directory

```
A quick google search suggested that I would need to run: `sudo apt-get update && sudo apt-get install libgl1-mesa-glx`. That seems to help, but then I get a different error message. 

```
/tmp/approxposterior/examples# python example.py 
Traceback (most recent call last):
  File "example.py", line 37, in <module>
    n_kl_samples=100000, verbose=True, debug=True)
  File "/root/anaconda3/lib/python3.6/site-packages/approxposterior/bp.py", line 247, in run
    save_plot="gp_mu_iter_%d.png" % nn)
  File "/root/anaconda3/lib/python3.6/site-packages/approxposterior/plot_utils.py", line 83, in plot_gp
    fig, ax = plt.subplots(**kw)
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/pyplot.py", line 1179, in subplots
    fig = figure(**fig_kw)
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/pyplot.py", line 534, in figure
    **kwargs)
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backend_bases.py", line 170, in new_figure_manager
    return cls.new_figure_manager_given_figure(num, fig)
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backend_bases.py", line 176, in new_figure_manager_given_figure
    canvas = cls.FigureCanvas(figure)
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backends/backend_qt5agg.py", line 35, in __init__
    super(FigureCanvasQTAggBase, self).__init__(figure=figure)
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backends/backend_qt5.py", line 235, in __init__
    _create_qApp()
  File "/root/anaconda3/lib/python3.6/site-packages/matplotlib/backends/backend_qt5.py", line 122, in _create_qApp
    raise RuntimeError('Invalid DISPLAY variable')
RuntimeError: Invalid DISPLAY variable
```

Thoughts? What am I missing now?

Under "A simple example", perhaps it would be useful if you could mention what the expected/correct output should look like. Also, does it make any difference if the code is run by Python 2 or 3? It is also worth mentioning in the readme.

Finally, "examples" link at https://github.com/dflemin3/approxposterior seems to be broken.

**Comment by dflemin3** [comment]
I'm not positive what's causing that bug, but I think it might be an issue with running matplotlib on a headless display.  I've tried adding `matplotlib.use('agg')` after importing matplotlib and before importing matplotlib.pyplot when I've encountered this before and this seems to fix it.  In this case, however, matplotlib.pyplot is implicitly imported in by plot_utils.py in bp.py, I'm not sure what the fix is and will need to dig a bit deeper.  @jakevdp any idea how to tackle this bug?

Note that even the simple example is computationally non-trivial and will take ~10s of minutes to run, so I've added that warning in the readme and the documentation.

Thanks for those suggestions!  I've added some code to examine the output and included the code and the figure it generates in the readme and online documentation to show what the example should produce.  The code is tested on Travis for 2.7+ and 3.6 so I've explicitly stated that in the readme.  Also, I corrected the broken link to the examples Jupyter Notebook section.

**Comment by jakevdp** [comment]
One approach would be to make matplotlib a soft requirement – only import it from within the functions where it is used. Presumably you should be able to use the bulk of the package even if matplotlib is not installed, yes?

**Comment by dflemin3** [comment]
I'm following @jakevdp suggestion since the plotting functions are for debugging purposes and the core functionality of approxposterior does not require matplotlib.  The master branch and pip versions have the change, and now I'm working on adding it to the conda-forge recipe.

**Comment by dflemin3** [comment]
The new release is v0.1.post1 and is up on pip.

**Comment by dmdu** [comment]
Sounds like good progress. I will take a closer look later today and tomorrow.

Is conda-forge updated now?

**Comment by dmdu** [comment]
pip installs the updated version and examples/example.py starts to run now. It outputs "X/10000" and repeats the sequence multiple times -- would it be possible to briefly describe in the README what these numbers and iterations mean?  Also, perhaps, the basic demo example should run quicker (maybe reduce the number of steps)? Then, would it be possible to add to your CI to always run/test this shorter example?

 Finally, are you planning to make `conda install -c conda-forge approxposterior` install the updated version?

**Comment by dflemin3** [comment]
I’m glad the pip version works now! That’s an excellent point, I do need to make a quicker example and briefly explain what that output is and how it can be silenced (verbose = False in the run method, for that case). It’ll be a bit tricky to make a quick example since the MCMC part of the code must run for a bit to draw enough samples from the posterior distribution, but I’ll think more carefully about that. I’m traveling until this Sunday and our internet is spotty at best, so I will address these points and work on updating the conda forge feed stock first thing next week.  Thank you again for the careful review!

**Comment by dmdu** [comment]
Sounds good!

**Comment by dflemin3** [comment]
The conda forge feed stock has been updated to the most recent release, v0.1.post1, and I've updated the example script with your suggestions.  I've shortened the example script by having the algorithm run for 2 main iterations, 5,000 MCMC iterations instead of 10,000, and I increased the initial training set size by an order of magnitude.  These changes still allow the algorithm to run through all core features, e.g. re-training the gaussian process surrogate model, and it still produces a good approximation to the true posterior distribution, as intended.  I have also silenced the X/10000 output by default and wrote a brief description on how one can turn on the output and what it means.  After these changes, the example script takes about 1 minute to run, so I've also included it as a new CI test and verified its accuracy.  I've updated both the README and github pages documentation to reflect these changes.
